runtime/src/snapshot_utils.rs

```diff
@@ -625,6 +627,76 @@ fn is_bank_snapshot_complete(bank_snapshot_dir: impl AsRef<Path>) -> bool {
     state_complete_path.is_file()
 }
 
+/// Writes the full snapshot slot file into the bank snapshot dir
+pub fn write_full_snapshot_slot_file(
+    bank_snapshot_dir: impl AsRef<Path>,
+    full_snapshot_slot: Slot,
+) -> IoResult<()> {
+    let full_snapshot_slot_path = bank_snapshot_dir
+        .as_ref()
+        .join(SNAPSHOT_FULL_SNAPSHOT_SLOT_FILENAME);
+    fs::write(
+        &full_snapshot_slot_path,
+        Slot::to_le_bytes(full_snapshot_slot),
+    )
+    .map_err(|err| {
+        IoError::other(format!(
+            "failed to write full snapshot slot file '{}': {err}",
+            full_snapshot_slot_path.display(),
+        ))
+    })
+}
+
+// Reads the full snapshot slot file from the bank snapshot dir
+pub fn read_full_snapshot_slot_file(bank_snapshot_dir: impl AsRef<Path>) -> IoResult<Slot> {
+    const SLOT_SIZE: usize = std::mem::size_of::<Slot>();
+    let full_snapshot_slot_path = bank_snapshot_dir
+        .as_ref()
+        .join(SNAPSHOT_FULL_SNAPSHOT_SLOT_FILENAME);
+    let full_snapshot_slot_file_metadata = fs::metadata(&full_snapshot_slot_path)?;
+    if full_snapshot_slot_file_metadata.len() != SLOT_SIZE as u64 {
+        let error_message = format!(
+            "invalid full snapshot slot file size: '{}' has {} bytes (should be {} bytes)",
+            full_snapshot_slot_path.display(),
+            full_snapshot_slot_file_metadata.len(),
+            SLOT_SIZE,
+        );
+        return Err(IoError::other(error_message));
+    }
+    let mut full_snapshot_slot_file = fs::File::open(&full_snapshot_slot_path)?;
+    let mut buffer = [0; SLOT_SIZE];
+    full_snapshot_slot_file.read_exact(&mut buffer)?;
+    let slot = Slot::from_le_bytes(buffer);
+    Ok(slot)
+}
+
+/// Gets the highest, loadable, bank snapshot
+///
+/// The highest bank snapshot is the one with the highest slot.
+/// To be loadable, the bank snapshot must be a BankSnapshotKind::Post.
+/// And if we're generating snapshots (e.g. running a normal validator), then
+/// the full snapshot file's slot must match the highest full snapshot archive's.
+pub fn get_highest_loadable_bank_snapshot(
```

Comment:
Had a separate discussion with @brooksprumo about this function.
There is an edge-case that is not handled as efficiently as we'd like, but we decided to leave it as a follow-up since this PR covers the majority of cases and already has a lot going on.

The edge case is when we have a higher full snapshot **archive** than our highest snapshot. This can happen if ledger-tool is used to create a full snapshot archive - this does not create a snapshot.

This function could likely be improved so that we only check the existence of the full snapshot archive matching the `full_snapshot_file_slot`, even if it not the latest full snapshot archive. However, it requires changes elsewhere in ABS (others?) to gate cleaning properly.